DataServer/portAllocator.py

```diff
@@ -28,22 +49,41 @@ def reservePort(self) -> int:
         Returns:
             int: Next available port number
         """
-        
-        # TODO reimplement this
-        self.__current += 1
-        if self.__current > self.__end:
-            raise RuntimeError
-        return self.__current
+        self.createDict(self.__start, self.__end)
+
+        for i in self.__portDict:
+            if self.__portDict[i]:
+                return_port = i
+                self.__s.bind((self.__ip, i))
+                self.__portDict[i] == False
+            else:
+                continue
+
+        if return_port > self.__end:
+            raise RuntimeError('All ports occupied')
+        return return_port
```

Comment:
If we can't find the port, `return_port` will never have been created.  `return_port` will then be accessed before being created.